### PR TITLE
Refactor: collated arch changes for naturality squares

### PR DESCRIPTION
**Pull request by regular-citizen**
This PR collects all architectural changes related to the implementation of naturality squares.

### CODE REVIEW

**Comment by thibautbenjamin** [comment]
I rebased this after merging the other branch

File: check_all.sh
Comment (thibautbenjamin) [inline check_all.sh:1]: Thank you so much for this, this is great. Could we maybe add it as a CI job to run this so we make sure that all test work ?

Comment (regular-citizen) [reply check_all.sh:1]: Implemented

**Comment by regular-citizen** [comment]
Ready for merging if no further reviews @thibautbenjamin

**Comment by thibautbenjamin** [comment]
This looks good to me, so I am merging it now

**Pull request merged**